Add doc for CA2020 (#31653)
dotnet/docs: docs/fundamentals/code-analysis/quality-rules/ca2020.md

f1_keywords:
  - PreventNumericIntPtrUIntPtrBehavioralChanges 
  - CA2020
helpviewer_keywords:
  - PreventNumericIntPtrUIntPtrBehavioralChanges 
  - CA2020
author: buyaa-n
---
# CA2020: Prevent behavioral change caused by built-in operators of IntPtr/UIntPtr

| | Value |
|-|-|
| **Rule ID** |CA2020|
| **Category** |[Reliability](reliability-warnings.md)|
| **Fix is breaking or non-breaking** |Non-breaking|

## Cause

This rule fires when it detects a behavioral change between .NET 6.0 and .NET 7.0 introduced with new built-in operators of numeric IntPtr/UIntPtr feature.

## Rule description

With the [numeric IntPtr feature](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/numeric-intptr.md), System.IntPtr and System.UIntPtr gained some built-in operators (conversions, unary and binary). Those now might throw when overflowing within checked context or may not throw in unchecked context compared to the previous `user-defined` operators in .NET 6.0 and below, which could cause behavioral change when upgrading from .NET 6.0 and below into .NET 7.

### List of APIs affected

| Operator | Context | In .NET 7 | In .NET 6 and earlier | Example |
|---|---|---|---|---|

> | operator +(IntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable + 2);` |
> | operator -(IntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable - 2);` |
> | explicit operator IntPtr(long) | Unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(IntPtr)longVariable;` |
> | explicit operator void\*(IntPtr) | Checked | throws when overflows | Doesn't throw when overflows | `checked((void*)intPtrVariable);` |
> | explicit operator IntPtr(void\*) | Checked | throws when overflows | Doesn't throw when overflows | `checked((IntPtr)voidPtrVariable);` |
> | explicit operator int(IntPtr) | Unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(int)intPtrVariable;` |
> | operator +(UIntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable + 2);` |
> | operator -(UIntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable - 2);` |
> | explicit operator UIntPtr(ulong) | Unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(UIntPtr)uLongVariable` |
> | explicit operator uint(UIntPtr) | Unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(uint)uintPtrVariable` |

Author: I think Checked/Unchecked better lower cased same as the checked/unchecked keywords

```suggestion
| operator +(IntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable + 2);` |
| operator -(IntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable - 2);` |
| explicit operator IntPtr(long) | unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(IntPtr)longVariable;` |
| explicit operator void\*(IntPtr) | checked | throws when overflows | Doesn't throw when overflows | `checked((void*)intPtrVariable);` |
| explicit operator IntPtr(void\*) | checked | throws when overflows | Doesn't throw when overflows | `checked((IntPtr)voidPtrVariable);` |
| explicit operator int(IntPtr) | unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(int)intPtrVariable;` |
| operator +(UIntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable + 2);` |
| operator -(UIntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable - 2);` |
| explicit operator UIntPtr(ulong) | unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(UIntPtr)uLongVariable` |
| explicit operator uint(UIntPtr) | unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(uint)uintPtrVariable` |
```

## How to fix violations

> Investigate the code to determine if the expression flagged could cause a behavioral change, and choose appropriate way to fix the diagnostic

Reviewer: ```suggestion
Examine your code to determine if the flagged expression could cause a behavioral change, and choose an appropriate way to fix the diagnostic from the following options:
```

Fix options:

- If the expression would not cause a behavioral change:

>   - If the `IntPtr/UIntPtr` type used as native int/uint purpose change the type to `nint/nuint`.

Reviewer: ```suggestion
  - If the `IntPtr` or `UIntPtr` type is used as a native `int` or `uint`, change the type to `nint` or `nuint`.
```

>   - If the `IntPtr/UIntPtr` type used as native pointer change the type to corresponding native pointer type

Reviewer: ```suggestion
  - If the `IntPtr` or `UIntPtr` type is used as a native pointer, change the type to the corresponding native pointer type.
```

>   - If you cannot change the type of the variable last resort will be suppress the warning

Reviewer: ```suggestion
  - If you can't change the type of the variable, suppress the warning.
```

> - If the expression could cause a behavioral change wrap it with corresponding `checked`/`unchecked` statement to preserve the .NET 6.0 behavior.

Reviewer: ```suggestion
- If the expression could cause a behavioral change, wrap it with a `checked` or `unchecked` statement to preserve the previous behavior.
```

            void* voidPtrVariable = (void*)intPtrVariable; // Starting with .NET 7 the explicit conversion '(void*)IntPtr' will throw when overflowing in a checked context. Wrap the expression with an 'unchecked' statement to restore the .NET 6 behavior.

            result = (IntPtr)voidPtrVariable; // Starting with .NET 7 the explicit conversion '(IntPtr)void*' will throw when overflowing in a checked context. Wrap the expression with an 'unchecked' statement to restore the .NET 6 behavior.
        }

        intPtrVariable = (IntPtr)longVariable; // Starting with .NET 7 the explicit conversion '(IntPtr)Int64' will not throw when overflowing in an unchecked context. Wrap the expression with a 'checked' statement to restore the .NET 6 behavior.

        int a = (int)intPtrVariable; // Starting with .NET 7 the explicit conversion '(Int32)IntPtr' will not throw when overflowing in an unchecked context. Wrap the expression with a 'checked' statement to restore the .NET 6 behavior.
    }
}
```

**Fix**:

> - If the expression would not cause a behavioral change:

Reviewer: ```suggestion
- If the expression would not cause a behavioral change and the the `IntPtr` or `UIntPtr` type is used as a native `int` or `uint`, change the type to `nint` or `nuint`.
```

>   - If the IntPtr/UIntPtr type used as native int/uint purpose change the type to nint/nuint.

Reviewer: ```suggestion
```

  ```csharp
  using System;

  public unsafe class IntPtrTest
  {
      nint intPtrVariable; // type changed to nint
      long longVariable;

      void Test ()
      {
          checked
          {
              nint result = intPtrVariable + 2; // no warning

              result = intPtrVariable - 2;
 
              void* voidPtrVariable = (void*)intPtrVariable;

              result = (nint)voidPtrVariable;
          }

          intPtrVariable = (nint)longVariable; 

          int a = (int)intPtrVariable;
      }
  }
  ```

> - If the expression could cause a behavioral change wrap it with corresponding `checked`/`unchecked` statement to preserve the old .NET 6.0 behavior.

Reviewer: ```suggestion
- If the expression could cause a behavioral change, wrap it with a `checked` or `unchecked` statement to preserve the previous behavior.
```

  ```csharp
  using System;

  public unsafe class IntPtrTest
  {
      IntPtr intPtrVariable;
      long longVariable;

      void Test ()
      {
          checked
          {
              IntPtr result = unchecked(intPtrVariable + 2); // wrap with unchecked

              result = unchecked(intPtrVariable - 2);

              void* voidPtrVariable = unchecked((void*)intPtrVariable); 

              result = unchecked((IntPtr)voidPtrVariable); 
          }

          intPtrVariable = checked((IntPtr)longVariable); // wrap with checked

          int a = checked((int)intPtrVariable); //
      }
  }
  ```

## When to suppress warnings

> If the expression would not cause a behavioral change it is safe to suppress a warning from this rule.

Reviewer: ```suggestion
If the expression would not cause a behavioral change, it's safe to suppress a warning from this rule.
```
